presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuilder.java

```diff
@@ -270,7 +271,10 @@ public void addInput(Page page)
                 size += index.getEstimatedSize().toBytes();
             }
 
-            operatorContext.setMemoryReservation(size);
+            // PagesIndex.addPage doesn't track page memory but blocks are shared and kept
+            pagesMemorySize += page.getRetainedSizeInBytes();
```

Comment:
As the comment, https://github.com/prestodb/presto/blob/c2b81c4858458e1c7e245880b989e40664249891/presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java#L145, says, pages memory is not counted. So it should be counted before the `addPage` returned. Otherwise it would raise OOM or cause full GC.

You're right about the race condition and we should address it. if I understand correctly the race condition has been existed regardless of this PR. `ParallelHashBuilderOperator` uses fresh new `OperatorContext` so there's no previous memory usage information in it. Ideally, `OperatorContext` memory should be transferred at `ParallelHashCollectOperator.finish` and `ParallelHashBuilderOperator` might need to get `SharedLookupSource` Future instead of `PagesIndex` Future. 

I'm going to add another commit about it.